symbolic-base/src/ZkFold/Symbolic/Data/List.hs

```diff
@@ -1,42 +1,158 @@
+
+uncons
+    :: forall context x
+    .  Symbolic context
+    => SymbolicData x
+    => Layout x ~ Vector (TypeSize x)
+    => Context x ~ context
+    => Context (List context x) ~ context
+    => Support x ~ Proxy context
+    => KnownNat (TypeSize x)
+    => List context x
+    -> (x, List context x)
+uncons l = (head l, tail l)
+
+-- | TODO: Is there really a nicer way to handle empty lists?
+--
+head
+    :: forall context x
+    .  Symbolic context
+    => SymbolicData x
+    => Layout x ~ Vector (TypeSize x)
+    => Context x ~ context
+    => Support x ~ Proxy context
+    => KnownNat (TypeSize x)
+    => List context x -> x
+head xs@List{..} = bool (restore $ \_ -> unsafeHead) (restore $ \_ -> embed $ pure zero) (null xs)
+    where
+        xRepr :: context (Vector (TypeSize x))
+        xRepr = case lWitness of
+                  (x:_) -> pieces x Proxy
+                  _     -> embed $ pure zero
+
+        -- | Head is a circuit comprised of variables satisfying the equation for prepending (i.e. (.:))
+        -- We have to pass witnesses to it as well.
+        --
+        unsafeHead :: context (Vector (TypeSize x))
+        unsafeHead = fromCircuit3F lHash xRepr lSize $
+            \vHash vRepr (Par1 s) -> V.zipWithM (\h r -> newConstrained (\p v -> p h + p v * (p s + one)) ($ r)) vHash vRepr
+
+tail
+    :: forall context x
+    .  Symbolic context
+    => SymbolicData x
+    => Layout x ~ Vector (TypeSize x)
+    => Layout (List context x) ~ Vector (TypeSize x + 1)
+    => Context x ~ context
+    => Context (List context x) ~ context
+    => Support x ~ Proxy context
+    => KnownNat (TypeSize x)
+    => List context x
+    -> List context x
+tail xs@List{..} = bool unsafeTail xs (null xs)
+    where
+        tailId :: [a] -> [a]
+        tailId []     = []
+        tailId (_:xs) = xs
 
--- TODO: we should implement an optimized version of `null`
+        unsafeTail :: List context x
+        unsafeTail = List decSum decSize (tailId lWitness)
 
-infixr 5 .:
-(.:) :: x -> List context x -> List context x
-_ .: _ = undefined
+        xRepr :: context (Vector (TypeSize x))
+        xRepr = case lWitness of
+                  (x:_) -> pieces x Proxy
+                  _     -> embed $ pure zero
 
-uncons :: List context x -> (x, List context x)
-uncons = undefined
+        decSum :: context (Vector (TypeSize x))
+        decSum = fromCircuit3F lHash xRepr lSize $
+            \vHash vRepr (Par1 s) -> V.zipWithM (\h r -> newAssigned (\p -> p h - p r * p s)) vHash vRepr
 
-head :: List context x -> x
-head xs = let (x, _) = uncons xs in x
+        decSize :: context Par1
+        decSize = fromCircuitF lSize (\(Par1 s) -> Par1 <$> newAssigned (\p -> p s - one))
 
-tail :: List context x -> List context x
-tail xs = let (_, xs') = uncons xs in xs'
+{-- TODO: Foldable relies on Haskell types such as Bool (i.e. null = foldr (\_ _ -> False) True)
```

Comment:
I think it shouldn't be Haskell's `Foldable`, but rather some similar typeclass which uses IVC inside